rviz_default_plugins/src/rviz_default_plugins/displays/depth_cloud/depth_cloud_display.cpp

```diff
@@ -0,0 +1,602 @@
+    delete ml_depth_data_;
+  }
+}
+
+void DepthCloudDisplay::setTopic(const QString & topic, const QString & datatype)
+{
+  // Copied from ImageDisplayBase::setTopic()
+  if (datatype == "sensor_msgs::msgs::Image") {
+    depth_transport_property_->setStdString("raw");
+    depth_topic_property_->setString(topic);
+  } else {
+    int index = topic.lastIndexOf("/");
+    if (index == -1) {
+      return;
+    }
+    QString transport = topic.mid(index + 1);
+    QString base_topic = topic.mid(0, index);
+
+    depth_transport_property_->setString(transport);
+    depth_topic_property_->setString(base_topic);
+  }
+}
+
+void DepthCloudDisplay::updateQueueSize()
+{
+  queue_size_ = queue_size_property_->getInt();
+}
+
+void DepthCloudDisplay::updateUseAutoSize()
+{
+  bool use_auto_size = use_auto_size_property_->getBool();
+  pointcloud_common_->point_world_size_property_->setReadOnly(use_auto_size);
+  pointcloud_common_->setAutoSize(use_auto_size);
+  auto_size_factor_property_->setHidden(!use_auto_size);
+  if (use_auto_size) {
+    use_auto_size_property_->expand();
+  }
+}
+
+void DepthCloudDisplay::updateAutoSizeFactor()
+{
+}
+
+void DepthCloudDisplay::updateTopicFilter()
+{
+  bool enabled = topic_filter_property_->getValue().toBool();
+  depth_topic_property_->enableFilter(enabled);
+  color_topic_property_->enableFilter(enabled);
+}
+
+void DepthCloudDisplay::updateUseOcclusionCompensation()
+{
+  bool use_occlusion_compensation = use_occlusion_compensation_property_->getBool();
+  occlusion_shadow_timeout_property_->setHidden(!use_occlusion_compensation);
+
+  if (use_occlusion_compensation) {
+    updateOcclusionTimeOut();
+    ml_depth_data_->enableOcclusionCompensation(true);
+    use_occlusion_compensation_property_->expand();
+  } else {
+    ml_depth_data_->enableOcclusionCompensation(false);
+  }
+}
+
+void DepthCloudDisplay::updateOcclusionTimeOut()
+{
+  float occlusion_timeout = occlusion_shadow_timeout_property_->getFloat();
+  ml_depth_data_->setShadowTimeOut(occlusion_timeout);
+}
+
+void DepthCloudDisplay::onEnable()
+{
+  subscribe();
+}
+
+void DepthCloudDisplay::onDisable()
+{
+  unsubscribe();
+
+  ml_depth_data_->reset();
+
+  clear();
+}
+
+void DepthCloudDisplay::subscribe()
+{
+  if (!isEnabled()) {
+    return;
+  }
+
+  try {
+    // reset all message filters
+    sync_depth_color_.reset(new SynchronizerDepthColor(SyncPolicyDepthColor(queue_size_)));
+    depthmap_tf_filter_.reset();
+    depthmap_sub_.reset(new image_transport::SubscriberFilter());
+    rgb_sub_.reset(new image_transport::SubscriberFilter());
+    cam_info_sub_.reset();
+
+    std::string depthmap_topic = depth_topic_property_->getTopicStd();
+    std::string color_topic = color_topic_property_->getTopicStd();
+
+    std::string depthmap_transport = depth_transport_property_->getStdString();
+    std::string color_transport = color_transport_property_->getStdString();
+
+    auto rviz_ros_node_ = context_->getRosNodeAbstraction();
```

Comment:
As with my comment above, I think we should just lock this right up front so we can check for an expired pointer and so that the rest of the function consistently has the object without it going away.